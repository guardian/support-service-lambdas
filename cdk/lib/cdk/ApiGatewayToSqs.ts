import { AwsIntegration, RestApi } from 'aws-cdk-lib/aws-apigateway';
import { Role, ServicePrincipal } from 'aws-cdk-lib/aws-iam';
import type { Queue } from 'aws-cdk-lib/aws-sqs';
import { Construct } from 'constructs';
import { SrApiGateway5xxAlarm } from './SrApiGateway5xxAlarm';
import type { SrMonitoring } from './SrLambdaAlarm';
import type { SrStack } from './SrStack';

type ApiGatewayToSqsProps = {
	queue: Queue;
	/**
	 * header names to pull from the request into SQS attributes.
	 *
	 * FIXME if a header is missing, the requester still gets a 200 and the message is silently dropped!
	 */
	includeHeaderNames: string[];
	/**
	 * do we want to disable standard SrCDK 5xx alarm or override any properties?
	 */
	monitoring: SrMonitoring;
};

/**
 * this takes incoming http requests and puts them on a queue, responding with a 200 OK
 *
 * Useful for webhooks etc.
 */
export class ApiGatewayToSqs extends Construct {
	constructor(scope: SrStack, id: string, props: ApiGatewayToSqsProps) {
		super(scope, id);

		const apiRole = new Role(this, 'Role', {
			assumedBy: new ServicePrincipal('apigateway.amazonaws.com'),
		});
		props.queue.grantSendMessages(apiRole);

		const sendMessageIntegration = new AwsIntegration({
			service: 'sqs',
			path: `${scope.account}/${props.queue.queueName}`,
			integrationHttpMethod: 'POST',
			options: {
				credentialsRole: apiRole,
				requestParameters: {
					'integration.request.header.Content-Type':
						"'application/x-www-form-urlencoded'",
				},
				requestTemplates: {
					'application/json': [
						...Object.entries({
							Action: 'SendMessage',
							MessageBody: '$util.urlEncode($input.body)',
						}),
						...props.includeHeaderNames.flatMap((headerName, index) => [
							[`MessageAttribute.${index + 1}.Name`, headerName],
							[`MessageAttribute.${index + 1}.Value.DataType`, 'String'],
							[
								`MessageAttribute.${index + 1}.Value.StringValue`,
								`$method.request.header.${headerName}`,
							],
						]),
					]
						.map(([k, v]) => k + '=' + v)
						.join('&'),
				},
				integrationResponses: [
					{
						statusCode: '200',
						responseTemplates: {
							'application/json': '{ "status": "accepted" }',
						},
					},
				],
			},
		});

		const apiGateway = new RestApi(scope, 'RestApi', {
			restApiName: `${scope.stack}-${scope.stage}-${scope.app}`,
		});

		if (scope.stage === 'PROD' && !props.monitoring.noMonitoring) {
			new SrApiGateway5xxAlarm(scope, {
				errorImpact: props.monitoring.errorImpact,
				restApi: apiGateway,
				lambdaFunctionNames: [],
			});
		}

		apiGateway.root
			.resourceForPath('/')
			.addMethod('POST', sendMessageIntegration, {
				methodResponses: [
					{
						statusCode: '200',
					},
				],
			});
	}
}
