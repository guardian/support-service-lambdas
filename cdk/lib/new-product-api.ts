import {join} from "path";
import {GuApiGatewayWithLambdaByPath} from "@guardian/cdk";
import {GuAlarm} from "@guardian/cdk/lib/constructs/cloudwatch";
import type {GuStackProps} from "@guardian/cdk/lib/constructs/core";
import {GuStack} from "@guardian/cdk/lib/constructs/core";
import {GuLambdaFunction} from "@guardian/cdk/lib/constructs/lambda";
import type {App} from "aws-cdk-lib";
import {Duration, Fn} from "aws-cdk-lib";
import {CfnBasePathMapping, CfnDomainName, Cors} from "aws-cdk-lib/aws-apigateway";
import {ComparisonOperator, Metric} from "aws-cdk-lib/aws-cloudwatch";
import {Effect, Policy, PolicyStatement} from "aws-cdk-lib/aws-iam";
import {Runtime} from "aws-cdk-lib/aws-lambda";
import {CfnRecordSet} from "aws-cdk-lib/aws-route53";
import {CfnInclude} from "aws-cdk-lib/cloudformation-include";

export interface NewProductApiProps extends GuStackProps {
  domainName: string;
  hostedZoneId: string;
  certificateId: string;
  apiGatewayTargetDomainName: string;
  zuoraCatalogLocation: string;
  fulfilmentDateCalculatorS3Resource: string;
}

export class NewProductApi extends GuStack {
  constructor(scope: App, id: string, props: NewProductApiProps) {
    super(scope, id, props);


    // ---- CFN template resources ---- //
    const yamlTemplateFilePath = join(__dirname, "../..", "handlers/new-product-api/cfn.yaml");
    new CfnInclude(this, "YamlTemplate", {
      templateFile: yamlTemplateFilePath,
    });


    // ---- Miscellaneous constants ---- //
    const isProd = this.stage === 'PROD';
    const app = "new-product-api";
    const runtime = Runtime.JAVA_11;
    const fileName = "new-product-api.jar";
    const memorySize = 1536;
    const timeout = Duration.seconds(300);
    const environment = {
      "Stage": this.stage,
      "EmailQueueName": Fn.importValue(`comms-${this.stage}-EmailQueueName`)
    };
    const sharedLambdaProps = {
      app,
      runtime,
      fileName,
      memorySize,
      timeout,
      environment,
    };


    // ---- API-triggered lambda functions ---- //
    const addSubscriptionLambda = new GuLambdaFunction(this, "add-subscription-cdk", {
      handler: "com.gu.newproduct.api.addsubscription.Handler::apply",
      functionName: `add-subscription-cdk-${this.stage}`,
      ...sharedLambdaProps,
    });

    const productCatalogLambda = new GuLambdaFunction(this, "product-catalog-cdk", {
      handler: "com.gu.newproduct.api.productcatalog.Handler::apply",
      functionName: `product-catalog-cdk-${this.stage}`,
      ...sharedLambdaProps,
    });


    // ---- API gateway ---- //
    const newProductApi = new GuApiGatewayWithLambdaByPath(this, {
      app,
      defaultCorsPreflightOptions: {
        allowOrigins: Cors.ALL_ORIGINS,
        allowMethods: Cors.ALL_METHODS,
        allowHeaders: ["Content-Type"],
      },
      monitoringConfiguration: {
        snsTopicName: "retention-dev",
        http5xxAlarm: {
          tolerated5xxPercentage: 1,
        }
      },
      targets: [
        {
          path: "/add-subscription",
          httpMethod: "POST",
          lambda: addSubscriptionLambda,
        },
        {
          path: "/product-catalog",
          httpMethod: "GET",
          lambda: productCatalogLambda,
        },
      ],
    })


    // ---- Alarms ---- //
    new GuAlarm(this, 'ApiGateway4XXAlarm', {
      app,
      alarmName: `new-product-api-${this.stage} API gateway 4XX response`,
      alarmDescription: "New Product API received an invalid request",
      evaluationPeriods: 1,
      threshold: 6,
      actionsEnabled: isProd,
      snsTopicName: "retention-dev",
      comparisonOperator: ComparisonOperator.GREATER_THAN_OR_EQUAL_TO_THRESHOLD,
      metric: new Metric({
        metricName: "4XXError",
        namespace: "AWS/ApiGateway",
        statistic: "Sum",
        period: Duration.seconds(900),
        dimensionsMap: {
          ApiName: `support-reminders-${this.stage}`,
        }
      }),
    });


    // ---- DNS ---- //
    // const certificateArn = `arn:aws:acm:${this.region}:${this.account}:certificate/${props.certificateId}`;
    //
    // const cfnDomainName = new CfnDomainName(this, "NewProductDomainName", {
    //   domainName: props.domainName,
    //   regionalCertificateArn: certificateArn,
    //   endpointConfiguration: {
    //     types: ["REGIONAL"]
    //   }
    // });
    //
    // new CfnBasePathMapping(this, "NewProductBasePathMapping", {
    //   domainName: cfnDomainName.ref,
    //   restApiId: newProductApi.api.restApiId,
    //   stage: newProductApi.api.deploymentStage.stageName,
    // });
    //
    // new CfnRecordSet(this, "NewProductDNSRecord", {
    //   name: props.domainName,
    //   type: "CNAME",
    //   hostedZoneId: props.hostedZoneId,
    //   ttl: "120",
    //   resourceRecords: [
    //     cfnDomainName.attrRegionalDomainName
    //   ],
    // });


    // ---- Apply policies ---- //
    const cloudwatchLogsInlinePolicy = (lambda: GuLambdaFunction, idPrefix: string): Policy => {
      return new Policy(this, `${idPrefix}-cloudwatch-logs-inline-policy`, {
        statements: [
          new PolicyStatement({
            effect: Effect.ALLOW,
            actions: [
              "logs:CreateLogGroup",
              "logs:CreateLogStream",
              "logs:PutLogEvents"
            ],
            resources: [
              `arn:aws:logs:${this.region}:${this.account}:log-group:/aws/lambda/${lambda.functionName}:log-stream:*`
            ]
          }),
        ],
      })
    }

    const addSubscriptionS3InlinePolicy: Policy = new Policy(this, "add-subscription-s3-inline-policy", {
      statements: [
        new PolicyStatement({
          effect: Effect.ALLOW,
          actions: [
            "s3:GetObject"
          ],
          resources: [
            `arn:aws:s3:::gu-reader-revenue-private/membership/support-service-lambdas/${this.stage}/zuoraRest-${this.stage}*.json`,
          ]
        }),
      ],
    })

    const sharedS3InlinePolicy: Policy = new Policy(this, "shared-s3-inline-policy", {
      statements: [
        new PolicyStatement({
          effect: Effect.ALLOW,
          actions: [
            "s3:GetObject"
          ],
          resources: [
            `arn:aws:s3:::fulfilment-date-calculator-${this.stage.toLowerCase()}/*`,
            `arn:aws:s3:::gu-zuora-catalog/${this.stage === "DEV" ? "CODE" : this.stage}/Zuora-${this.stage}/catalog.json`
          ]
        }),
      ],
    })

    const sqsInlinePolicy: Policy = new Policy(this, "sqs-inline-policy", {
      statements: [
        new PolicyStatement({
          effect: Effect.ALLOW,
          actions: [
            "sqs:GetQueueUrl",
            "sqs:SendMessage"
          ],
          resources: [
            `arn:aws:sqs:${this.region}:${this.account}:braze-emails-${this.stage}`
          ]
        }),
      ],
    })

    addSubscriptionLambda.role?.attachInlinePolicy(cloudwatchLogsInlinePolicy(addSubscriptionLambda, "add-subscription"))
    addSubscriptionLambda.role?.attachInlinePolicy(sharedS3InlinePolicy)
    addSubscriptionLambda.role?.attachInlinePolicy(addSubscriptionS3InlinePolicy)
    addSubscriptionLambda.role?.attachInlinePolicy(sqsInlinePolicy)

    productCatalogLambda.role?.attachInlinePolicy(cloudwatchLogsInlinePolicy(productCatalogLambda, "product-catalog"))
    productCatalogLambda.role?.attachInlinePolicy(sharedS3InlinePolicy)
  }
}
