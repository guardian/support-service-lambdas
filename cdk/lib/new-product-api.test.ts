import { App } from 'aws-cdk-lib';
import { Template } from 'aws-cdk-lib/assertions';
import { codeProps, prodProps } from '../bin/cdk';
import { NewProductApi } from './new-product-api';

describe('The NewProductApi stack', () => {
	it('matches the snapshot', () => {
		const app = new App();
		const codeStack = new NewProductApi(app, 'new-product-api-CODE', codeProps);
		const prodStack = new NewProductApi(app, 'new-product-api-PROD', prodProps);
		expect(Template.fromStack(codeStack).toJSON()).toMatchSnapshot();
		expect(Template.fromStack(prodStack).toJSON()).toMatchSnapshot();
	});
});
