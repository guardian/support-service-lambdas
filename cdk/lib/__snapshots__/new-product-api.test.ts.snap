// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`The NewProductApi stack matches the snapshot 1`] = `
{
  "AWSTemplateFormatVersion": "2010-09-09",
  "Conditions": {
    "CreateProdMonitoring": {
      "Fn::Equals": [
        {
          "Ref": "Stage",
        },
        "PROD",
      ],
    },
  },
  "Description": "new product api",
  "Mappings": {
    "StageMap": {
      "CODE": {
        "ApiGatewayTargetDomainName": "d-ecyddyj7nk.execute-api.eu-west-1.amazonaws.com",
        "ApiName": "new-product-api-CODE",
        "DomainName": "new-product-api-code.membership.guardianapis.com",
        "FulfilmentDateCalculatorS3Resource": "arn:aws:s3:::fulfilment-date-calculator-code/*",
        "ZuoraCatalogLocation": "arn:aws:s3:::gu-zuora-catalog/CODE/Zuora-UAT/catalog.json",
      },
      "DEV": {
        "ApiGatewayTargetDomainName": "d-papcvkgmc9.execute-api.eu-west-1.amazonaws.com",
        "ApiName": "new-product-api-DEV",
        "DomainName": "new-product-api-dev.membership.guardianapis.com",
        "FulfilmentDateCalculatorS3Resource": "arn:aws:s3:::fulfilment-date-calculator-dev/*",
        "ZuoraCatalogLocation": "arn:aws:s3:::gu-zuora-catalog/CODE/Zuora-DEV/catalog.json",
      },
      "PROD": {
        "ApiGatewayTargetDomainName": "d-yyh9pmqphi.execute-api.eu-west-1.amazonaws.com",
        "ApiName": "new-product-api-PROD",
        "DomainName": "new-product-api-prod.membership.guardianapis.com",
        "FulfilmentDateCalculatorS3Resource": "arn:aws:s3:::fulfilment-date-calculator-prod/*",
        "ZuoraCatalogLocation": "arn:aws:s3:::gu-zuora-catalog/PROD/Zuora-PROD/catalog.json",
      },
    },
  },
  "Metadata": {
    "gu:cdk:constructs": [],
    "gu:cdk:version": "TEST",
  },
  "Parameters": {
    "Stage": {
      "AllowedValues": [
        "PROD",
        "CODE",
        "DEV",
      ],
      "Default": "CODE",
      "Description": "Stage name",
      "Type": "String",
    },
  },
  "Resources": {
    "4xxApiAlarm": {
      "Condition": "CreateProdMonitoring",
      "Properties": {
        "AlarmActions": [
          {
            "Fn::Sub": "arn:aws:sns:\${AWS::Region}:\${AWS::AccountId}:retention-dev",
          },
        ],
        "AlarmName": {
          "Fn::Sub": [
            "4XX rate from \${ApiName}",
            {
              "ApiName": {
                "Fn::FindInMap": [
                  "StageMap",
                  {
                    "Ref": "Stage",
                  },
                  "ApiName",
                ],
              },
            },
          ],
        },
        "ComparisonOperator": "GreaterThanThreshold",
        "Dimensions": [
          {
            "Name": "ApiName",
            "Value": {
              "Fn::FindInMap": [
                "StageMap",
                {
                  "Ref": "Stage",
                },
                "ApiName",
              ],
            },
          },
          {
            "Name": "Stage",
            "Value": {
              "Fn::Sub": "\${Stage}",
            },
          },
        ],
        "EvaluationPeriods": 1,
        "MetricName": "4XXError",
        "Namespace": "AWS/ApiGateway",
        "Period": 900,
        "Statistic": "Sum",
        "Threshold": 6,
        "TreatMissingData": "notBreaching",
      },
      "Type": "AWS::CloudWatch::Alarm",
    },
    "5xxApiAlarm": {
      "Condition": "CreateProdMonitoring",
      "Properties": {
        "AlarmActions": [
          {
            "Fn::Sub": "arn:aws:sns:\${AWS::Region}:\${AWS::AccountId}:retention-dev",
          },
        ],
        "AlarmName": {
          "Fn::Sub": [
            "5XX rate from \${ApiName}",
            {
              "ApiName": {
                "Fn::FindInMap": [
                  "StageMap",
                  {
                    "Ref": "Stage",
                  },
                  "ApiName",
                ],
              },
            },
          ],
        },
        "ComparisonOperator": "GreaterThanThreshold",
        "Dimensions": [
          {
            "Name": "ApiName",
            "Value": {
              "Fn::FindInMap": [
                "StageMap",
                {
                  "Ref": "Stage",
                },
                "ApiName",
              ],
            },
          },
          {
            "Name": "Stage",
            "Value": {
              "Fn::Sub": "\${Stage}",
            },
          },
        ],
        "EvaluationPeriods": 1,
        "MetricName": "5XXError",
        "Namespace": "AWS/ApiGateway",
        "Period": 900,
        "Statistic": "Sum",
        "Threshold": 6,
        "TreatMissingData": "notBreaching",
      },
      "Type": "AWS::CloudWatch::Alarm",
    },
    "AddSubscriptionLambda": {
      "DependsOn": [
        "NewSubscriptionRole",
      ],
      "Properties": {
        "Code": {
          "S3Bucket": "support-service-lambdas-dist",
          "S3Key": {
            "Fn::Sub": "membership/\${Stage}/new-product-api/new-product-api.jar",
          },
        },
        "Description": "add subscription to account",
        "Environment": {
          "Variables": {
            "EmailQueueName": {
              "Fn::ImportValue": {
                "Fn::Sub": "comms-\${Stage}-EmailQueueName",
              },
            },
            "Stage": {
              "Ref": "Stage",
            },
          },
        },
        "FunctionName": {
          "Fn::Sub": "add-subscription-\${Stage}",
        },
        "Handler": "com.gu.newproduct.api.addsubscription.Handler::apply",
        "MemorySize": 1536,
        "Role": {
          "Fn::GetAtt": [
            "NewSubscriptionRole",
            "Arn",
          ],
        },
        "Runtime": "java11",
        "Tags": [
          {
            "Key": "gu:cdk:version",
            "Value": "TEST",
          },
          {
            "Key": "gu:repo",
            "Value": "guardian/support-service-lambdas",
          },
          {
            "Key": "Stack",
            "Value": "membership",
          },
          {
            "Key": "Stage",
            "Value": "TEST",
          },
        ],
        "Timeout": 300,
      },
      "Type": "AWS::Lambda::Function",
    },
    "AddSubscriptionMethod": {
      "DependsOn": [
        "AddSubscriptionLambda",
        "AddSubscriptionResource",
        "NewProductApi",
      ],
      "Properties": {
        "ApiKeyRequired": true,
        "AuthorizationType": "NONE",
        "HttpMethod": "POST",
        "Integration": {
          "IntegrationHttpMethod": "POST",
          "Type": "AWS_PROXY",
          "Uri": {
            "Fn::Sub": "arn:aws:apigateway:\${AWS::Region}:lambda:path/2015-03-31/functions/\${AddSubscriptionLambda.Arn}/invocations",
          },
        },
        "ResourceId": {
          "Ref": "AddSubscriptionResource",
        },
        "RestApiId": {
          "Ref": "NewProductApi",
        },
      },
      "Type": "AWS::ApiGateway::Method",
    },
    "AddSubscriptionResource": {
      "DependsOn": [
        "NewProductApi",
      ],
      "Properties": {
        "ParentId": {
          "Fn::GetAtt": [
            "NewProductApi",
            "RootResourceId",
          ],
        },
        "PathPart": "add-subscription",
        "RestApiId": {
          "Ref": "NewProductApi",
        },
      },
      "Type": "AWS::ApiGateway::Resource",
    },
    "NewProductAPIPermission": {
      "DependsOn": [
        "AddSubscriptionLambda",
      ],
      "Properties": {
        "Action": "lambda:invokeFunction",
        "FunctionName": {
          "Fn::Sub": "add-subscription-\${Stage}",
        },
        "Principal": "apigateway.amazonaws.com",
      },
      "Type": "AWS::Lambda::Permission",
    },
    "NewProductApi": {
      "Properties": {
        "Description": "salesforce calls this api to enable csrs to add subscriptions to accounts",
        "Name": {
          "Fn::FindInMap": [
            "StageMap",
            {
              "Ref": "Stage",
            },
            "ApiName",
          ],
        },
        "Tags": [
          {
            "Key": "gu:cdk:version",
            "Value": "TEST",
          },
          {
            "Key": "gu:repo",
            "Value": "guardian/support-service-lambdas",
          },
          {
            "Key": "Stack",
            "Value": "membership",
          },
          {
            "Key": "Stage",
            "Value": "TEST",
          },
        ],
      },
      "Type": "AWS::ApiGateway::RestApi",
    },
    "NewProductApiDeployment": {
      "DependsOn": [
        "AddSubscriptionMethod",
        "ProductCatalogMethod",
      ],
      "Properties": {
        "Description": "Deploys new-product-api into an environment/stage",
        "RestApiId": {
          "Ref": "NewProductApi",
        },
      },
      "Type": "AWS::ApiGateway::Deployment",
    },
    "NewProductApiKey": {
      "DependsOn": [
        "NewProductApi",
        "NewProductApiStage",
      ],
      "Properties": {
        "Description": "Key required to call new product API",
        "Enabled": true,
        "Name": {
          "Fn::Sub": "new-product-api-key-\${Stage}",
        },
        "StageKeys": [
          {
            "RestApiId": {
              "Ref": "NewProductApi",
            },
            "StageName": {
              "Fn::Sub": "\${Stage}",
            },
          },
        ],
        "Tags": [
          {
            "Key": "gu:cdk:version",
            "Value": "TEST",
          },
          {
            "Key": "gu:repo",
            "Value": "guardian/support-service-lambdas",
          },
          {
            "Key": "Stack",
            "Value": "membership",
          },
          {
            "Key": "Stage",
            "Value": "TEST",
          },
        ],
      },
      "Type": "AWS::ApiGateway::ApiKey",
    },
    "NewProductApiStage": {
      "DependsOn": [
        "AddSubscriptionMethod",
        "ProductCatalogMethod",
      ],
      "Properties": {
        "DeploymentId": {
          "Ref": "NewProductApiDeployment",
        },
        "Description": "Stage for new product api",
        "RestApiId": {
          "Ref": "NewProductApi",
        },
        "StageName": {
          "Fn::Sub": "\${Stage}",
        },
        "Tags": [
          {
            "Key": "gu:cdk:version",
            "Value": "TEST",
          },
          {
            "Key": "gu:repo",
            "Value": "guardian/support-service-lambdas",
          },
          {
            "Key": "Stack",
            "Value": "membership",
          },
          {
            "Key": "Stage",
            "Value": "TEST",
          },
        ],
      },
      "Type": "AWS::ApiGateway::Stage",
    },
    "NewProductBasePathMapping": {
      "DependsOn": [
        "NewProductApi",
        "NewProductApiStage",
        "NewProductDomainName",
      ],
      "Properties": {
        "DomainName": {
          "Ref": "NewProductDomainName",
        },
        "RestApiId": {
          "Ref": "NewProductApi",
        },
        "Stage": {
          "Fn::Sub": "\${Stage}",
        },
      },
      "Type": "AWS::ApiGateway::BasePathMapping",
    },
    "NewProductDNSRecord": {
      "Properties": {
        "Comment": {
          "Fn::Sub": "CNAME for new product API \${Stage}",
        },
        "HostedZoneName": "membership.guardianapis.com.",
        "Name": {
          "Fn::FindInMap": [
            "StageMap",
            {
              "Ref": "Stage",
            },
            "DomainName",
          ],
        },
        "ResourceRecords": [
          {
            "Fn::FindInMap": [
              "StageMap",
              {
                "Ref": "Stage",
              },
              "ApiGatewayTargetDomainName",
            ],
          },
        ],
        "TTL": "120",
        "Type": "CNAME",
      },
      "Type": "AWS::Route53::RecordSet",
    },
    "NewProductDomainName": {
      "Properties": {
        "DomainName": {
          "Fn::FindInMap": [
            "StageMap",
            {
              "Ref": "Stage",
            },
            "DomainName",
          ],
        },
        "EndpointConfiguration": {
          "Types": [
            "REGIONAL",
          ],
        },
        "RegionalCertificateArn": {
          "Fn::Sub": "arn:aws:acm:\${AWS::Region}:\${AWS::AccountId}:certificate/c1efc564-9ff8-4a03-be48-d1990a3d79d2",
        },
        "Tags": [
          {
            "Key": "gu:cdk:version",
            "Value": "TEST",
          },
          {
            "Key": "gu:repo",
            "Value": "guardian/support-service-lambdas",
          },
          {
            "Key": "Stack",
            "Value": "membership",
          },
          {
            "Key": "Stage",
            "Value": "TEST",
          },
        ],
      },
      "Type": "AWS::ApiGateway::DomainName",
    },
    "NewProductUsagePlan": {
      "DependsOn": [
        "NewProductApi",
        "NewProductApiStage",
      ],
      "Properties": {
        "ApiStages": [
          {
            "ApiId": {
              "Ref": "NewProductApi",
            },
            "Stage": {
              "Ref": "NewProductApiStage",
            },
          },
        ],
        "Tags": [
          {
            "Key": "gu:cdk:version",
            "Value": "TEST",
          },
          {
            "Key": "gu:repo",
            "Value": "guardian/support-service-lambdas",
          },
          {
            "Key": "Stack",
            "Value": "membership",
          },
          {
            "Key": "Stage",
            "Value": "TEST",
          },
        ],
        "UsagePlanName": {
          "Fn::Sub": "new-product-api-usage-plan-\${Stage}",
        },
      },
      "Type": "AWS::ApiGateway::UsagePlan",
    },
    "NewProductUsagePlanKey": {
      "DependsOn": [
        "NewProductApiKey",
        "NewProductUsagePlan",
      ],
      "Properties": {
        "KeyId": {
          "Ref": "NewProductApiKey",
        },
        "KeyType": "API_KEY",
        "UsagePlanId": {
          "Ref": "NewProductUsagePlan",
        },
      },
      "Type": "AWS::ApiGateway::UsagePlanKey",
    },
    "NewSubscriptionRole": {
      "Properties": {
        "AssumeRolePolicyDocument": {
          "Statement": [
            {
              "Action": [
                "sts:AssumeRole",
              ],
              "Effect": "Allow",
              "Principal": {
                "Service": [
                  "lambda.amazonaws.com",
                ],
              },
            },
          ],
        },
        "Path": "/",
        "Policies": [
          {
            "PolicyDocument": {
              "Statement": [
                {
                  "Action": [
                    "logs:CreateLogGroup",
                    "logs:CreateLogStream",
                    "logs:PutLogEvents",
                    "lambda:InvokeFunction",
                  ],
                  "Effect": "Allow",
                  "Resource": {
                    "Fn::Sub": "arn:aws:logs:\${AWS::Region}:\${AWS::AccountId}:log-group:/aws/lambda/add-subscription-\${Stage}:log-stream:*",
                  },
                },
              ],
            },
            "PolicyName": "LambdaPolicy",
          },
          {
            "PolicyDocument": {
              "Statement": [
                {
                  "Action": "s3:GetObject",
                  "Effect": "Allow",
                  "Resource": [
                    {
                      "Fn::Sub": "arn:aws:s3:::gu-reader-revenue-private/membership/support-service-lambdas/\${Stage}/zuoraRest-\${Stage}*.json",
                    },
                    {
                      "Fn::FindInMap": [
                        "StageMap",
                        {
                          "Ref": "Stage",
                        },
                        "FulfilmentDateCalculatorS3Resource",
                      ],
                    },
                  ],
                },
              ],
            },
            "PolicyName": "ReadPrivateCredentials",
          },
          {
            "PolicyDocument": {
              "Statement": [
                {
                  "Action": [
                    "sqs:GetQueueUrl",
                    "sqs:SendMessage",
                  ],
                  "Effect": "Allow",
                  "Resource": {
                    "Fn::ImportValue": {
                      "Fn::Sub": "comms-\${Stage}-EmailQueueArn",
                    },
                  },
                },
              ],
            },
            "PolicyName": "SQSAddEmailRequest",
          },
          {
            "PolicyDocument": {
              "Statement": [
                {
                  "Action": "s3:GetObject",
                  "Effect": "Allow",
                  "Resource": {
                    "Fn::FindInMap": [
                      "StageMap",
                      {
                        "Ref": "Stage",
                      },
                      "ZuoraCatalogLocation",
                    ],
                  },
                },
              ],
            },
            "PolicyName": "ReadZuoraCatalog",
          },
        ],
        "Tags": [
          {
            "Key": "gu:cdk:version",
            "Value": "TEST",
          },
          {
            "Key": "gu:repo",
            "Value": "guardian/support-service-lambdas",
          },
          {
            "Key": "Stack",
            "Value": "membership",
          },
          {
            "Key": "Stage",
            "Value": "TEST",
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "ProductCatalogApiPermission": {
      "DependsOn": [
        "ProductCatalogLambda",
      ],
      "Properties": {
        "Action": "lambda:invokeFunction",
        "FunctionName": {
          "Fn::Sub": "product-catalog-\${Stage}",
        },
        "Principal": "apigateway.amazonaws.com",
      },
      "Type": "AWS::Lambda::Permission",
    },
    "ProductCatalogLambda": {
      "DependsOn": [
        "ProductCatalogRole",
      ],
      "Properties": {
        "Code": {
          "S3Bucket": "support-service-lambdas-dist",
          "S3Key": {
            "Fn::Sub": "membership/\${Stage}/new-product-api/new-product-api.jar",
          },
        },
        "Description": "get description of available products",
        "Environment": {
          "Variables": {
            "Stage": {
              "Ref": "Stage",
            },
          },
        },
        "FunctionName": {
          "Fn::Sub": "product-catalog-\${Stage}",
        },
        "Handler": "com.gu.newproduct.api.productcatalog.Handler::apply",
        "MemorySize": 1536,
        "Role": {
          "Fn::GetAtt": [
            "ProductCatalogRole",
            "Arn",
          ],
        },
        "Runtime": "java11",
        "Tags": [
          {
            "Key": "gu:cdk:version",
            "Value": "TEST",
          },
          {
            "Key": "gu:repo",
            "Value": "guardian/support-service-lambdas",
          },
          {
            "Key": "Stack",
            "Value": "membership",
          },
          {
            "Key": "Stage",
            "Value": "TEST",
          },
        ],
        "Timeout": 300,
      },
      "Type": "AWS::Lambda::Function",
    },
    "ProductCatalogMethod": {
      "DependsOn": [
        "NewProductApi",
        "ProductCatalogLambda",
        "ProductCatalogResource",
      ],
      "Properties": {
        "ApiKeyRequired": true,
        "AuthorizationType": "NONE",
        "HttpMethod": "GET",
        "Integration": {
          "IntegrationHttpMethod": "POST",
          "Type": "AWS_PROXY",
          "Uri": {
            "Fn::Sub": "arn:aws:apigateway:\${AWS::Region}:lambda:path/2015-03-31/functions/\${ProductCatalogLambda.Arn}/invocations",
          },
        },
        "ResourceId": {
          "Ref": "ProductCatalogResource",
        },
        "RestApiId": {
          "Ref": "NewProductApi",
        },
      },
      "Type": "AWS::ApiGateway::Method",
    },
    "ProductCatalogResource": {
      "DependsOn": [
        "NewProductApi",
      ],
      "Properties": {
        "ParentId": {
          "Fn::GetAtt": [
            "NewProductApi",
            "RootResourceId",
          ],
        },
        "PathPart": "product-catalog",
        "RestApiId": {
          "Ref": "NewProductApi",
        },
      },
      "Type": "AWS::ApiGateway::Resource",
    },
    "ProductCatalogRole": {
      "Properties": {
        "AssumeRolePolicyDocument": {
          "Statement": [
            {
              "Action": [
                "sts:AssumeRole",
              ],
              "Effect": "Allow",
              "Principal": {
                "Service": [
                  "lambda.amazonaws.com",
                ],
              },
            },
          ],
        },
        "Path": "/",
        "Policies": [
          {
            "PolicyDocument": {
              "Statement": [
                {
                  "Action": [
                    "logs:CreateLogGroup",
                    "logs:CreateLogStream",
                    "logs:PutLogEvents",
                    "lambda:InvokeFunction",
                  ],
                  "Effect": "Allow",
                  "Resource": {
                    "Fn::Sub": "arn:aws:logs:\${AWS::Region}:\${AWS::AccountId}:log-group:/aws/lambda/product-catalog-\${Stage}:log-stream:*",
                  },
                },
              ],
            },
            "PolicyName": "LambdaPolicy",
          },
          {
            "PolicyDocument": {
              "Statement": [
                {
                  "Action": "s3:GetObject",
                  "Effect": "Allow",
                  "Resource": [
                    {
                      "Fn::FindInMap": [
                        "StageMap",
                        {
                          "Ref": "Stage",
                        },
                        "ZuoraCatalogLocation",
                      ],
                    },
                    {
                      "Fn::FindInMap": [
                        "StageMap",
                        {
                          "Ref": "Stage",
                        },
                        "FulfilmentDateCalculatorS3Resource",
                      ],
                    },
                  ],
                },
              ],
            },
            "PolicyName": "ReadZuoraCatalog",
          },
        ],
        "Tags": [
          {
            "Key": "gu:cdk:version",
            "Value": "TEST",
          },
          {
            "Key": "gu:repo",
            "Value": "guardian/support-service-lambdas",
          },
          {
            "Key": "Stack",
            "Value": "membership",
          },
          {
            "Key": "Stage",
            "Value": "TEST",
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
  },
}
`;
